# Домашнее задание 1
## Майнер Биоинформатика (2-й год)
### Кочеткова Ирина (группа 4)

## Работа на сервере
Список команд для сервера находится в файле README, файлы с отчетами и инфографикой выложены в репозитории.
В бонусной части задания брались 1000000 и 500000 чтений соответственно.

time platanus gap_close -o Poil -c Poil_small_scaffold.fa -IP1 small_sub1.fastq.trimmed small_sub2.fastq.trimmed -OP2 small_matep1.fastq.int_trimmed small_matep2.fastq.int_trimmed 2> gapclose.log

## Анализ данных

In [7]:
import re

def info(file, text, output = True):
    lens = []
    sum_len = 0
    num = 0
    max_len = 0
    length = 0
    score = 0
    max_sequence = ''
    sequence = ''
    for line in file:
        if (line[0] == '>'):
            if num != 0:
                lens.append(length)
            num += 1
            if length >= max_len:
                max_len = length
                max_sequence = sequence
            sequence = ''
            length = 0
        else:
            sequence += line.strip()
            length += len(line.strip())
            sum_len += len(line.strip())
     
    lens.sort(reverse = True) 
    for i in lens:
        score += i
        if score >= sum_len / 2:
            if output == True:
                print(f'Анализ {text}\n\
Количество: {num},\n\
Длина: {sum_len},\n\
Самый длинный: {max_len},\n\
N50: {i}\n')
            break
    return max_sequence

In [8]:
max_cont = info(open('Poil_contig.fa', 'r'), 'Контигов')

Анализ Контигов
Количество: 619,
Длина: 3924910,
Самый длинный: 179307,
N50: 52802



In [13]:
small_cont = info(open('Poil_small_contig.fa', 'r'), 'Контигов (маленькая последовательность)')

Анализ Контигов (маленькая последовательность)
Количество: 269,
Длина: 1030079,
Самый длинный: 187087,
N50: 64093



### Вывод для контигов:


*   С уменьшением чтений количество контигов меньше (619 vs. 269)
*   С уменьшением чтений N50 больше (52802 vs. 64093)
*   С уменьшением чтений общая длина уменьшается  (3924910 vs. 1030079)
*   Длина самого длинного не зависит от количества чтений



In [9]:
max_scaf = info(open('Poil_scaffold.fa', 'r'), 'Скаффолдов')

Анализ Скаффолдов
Количество: 70,
Длина: 3873294,
Самый длинный: 3834030,
N50: 3834030



In [16]:
small_scaf = info(open('Poil_small_scaffold.fa', 'r'), 'Скаффолдов (маленькая последовательность)')

Анализ Скаффолдов (маленькая последовательность)
Количество: 152,
Длина: 3854239,
Самый длинный: 2962909,
N50: 2962909



### Вывод для скаффолдов:


*   С уменьшением чтений количество скаффолдов больше (70 vs. 152)
*   С уменьшением чтений N50 меньше (3834030 vs. 2962909)
*   С уменьшением чтений общая длина уменьшается  (3873294 vs. 3854239)
*   Длина самого длинного не зависит от количества чтений

In [10]:
print(f'Длина пробелов: {max_scaf.count("N")}')
max_scaf = re.sub(r'N{2,}', 'N', max_scaf)
print(f'Кол-во пробелов: {max_scaf.count("N")}')

Длина пробелов: 7554
Кол-во пробелов: 65


In [17]:
print(f'Длина пробелов: {small_scaf.count("N")}')
small_scaf = re.sub(r'N{2,}', 'N', small_scaf)
print(f'Кол-во пробелов: {small_scaf.count("N")}')

Длина пробелов: 11182
Кол-во пробелов: 99




*   Количество пробелов и их общая длина увеличились



In [12]:
max_scaf = info(open('Poil_gapClosed.fa', 'r'), 'Скаффолдов', False)
print(f'Длина пробелов для обрезанных чтений: {max_scaf.count("N")}')
max_scaf = re.sub(r'N{2,}', 'N', max_scaf)
print(f'Кол-во пробелов для обрезанных чтений: {max_scaf.count("N")}')

Длина пробелов для обрезанных чтений: 2969
Кол-во пробелов для обрезанных чтений: 9


In [18]:
small_scaf = info(open('Poil_small_gapClosed.fa', 'r'), 'Скаффолдов (маленькая последовательность)', False)
print(f'Длина пробелов для обрезанных чтений: {small_scaf.count("N")}')
small_scaf = re.sub(r'N{2,}', 'N', small_scaf)
print(f'Кол-во пробелов для обрезанных чтений: {small_scaf.count("N")}')

Длина пробелов для обрезанных чтений: 5620
Кол-во пробелов для обрезанных чтений: 28


*   Количество пробелов и их общая длина увеличились